CMakeLists.txt

```diff
@@ -9,9 +9,15 @@ set(PROJECT_VERSION_MINOR 8)
 set(CMAKE_CXX_STANDARD 98)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
-set(CMAKE_BUILD_TYPE
-    RelWithDebInfo
-    CACHE STRING "Empty or one of Debug, Release, RelWithDebInfo")
+set(allowableBuildTypes Debug Release RelWithDebInfo)
+if(NOT CMAKE_BUILD_TYPE) 
+  set(CMAKE_BUILD_TYPE
+      RelWithDebInfo
+      CACHE STRING "Empty or one of ${allowableBuildTypes}" FORCE)
+elseif(NOT CMAKE_BUILD_TYPE IN_LIST allowableBuildTypes)
```

Comment:
Is there any reason not to allow some custom build type? For example, in CoreNEURON we [allowed that here](https://github.com/BlueBrain/CoreNeuron#optimization-flags).

What we have done in the past is that we have some helper CMake file like [CompilerFlagsHelpers.cmake](https://github.com/BlueBrain/CoreNeuron/blob/master/CMake/config/CompilerFlagsHelpers.cmake) that sets flags for various compilers. And then we use config file like [ReleaseDebugAutoFlags.cmake ](https://github.com/BlueBrain/CoreNeuron/blob/master/CMake/config/ReleaseDebugAutoFlags.cmake) to define flag for different CMAKE_BUILD_TYPE.

We can do it differently but I am pointing out the way that we have been using.

Reply:
That is a good idea. I'll follow the CoreNEURON approach. Should that be done now or does it overly complicate the review process and should be done later?

Reply:
Thanks @pramodk for pointing this out. I actually didn't know this part of CoreNEURON build system. @nrnhines, no I think you can commit into this PR, it seems to be like a sensible improvement.

Reply:
I added this as best I could but probably needs some line location changes. I modified the ReleaseDebugAutoFlags.cmake  to use my allowableBuildTypes list and CMAKE_BUILD_TYPE_DEFAULT . However, because this was near the beginning of CMakeLists.txt, I had to miove the CMAKE_MODULE_PATH list from its old location further
down the file.  If the CMAKE_BUILD_TYPE does not have to be near the beginning...

By the way the CoreNEURON CMakeLists.txt near the beginning has what seems to be an
obsolete fragment:
```
set(CMAKE_BUILD_TYPE
        RelWithDebInfo
        CACHE STRING "Empty or one of Debug, Release, RelWithDebInfo")
```

Reply:
@nrnhines : ok. above coreneuron related change we can address separately.